cns/deviceplugin/server.go

```diff
@@ -0,0 +1,153 @@
+package deviceplugin
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1"
+)
+
+type deviceCounter interface {
+	getDeviceCount() int
+}
+
+type Server struct {
+	address              string
+	logger               *zap.Logger
+	deviceCounter        deviceCounter
+	listAndWatchDoneChan chan struct{}
+	deviceCheckInterval  time.Duration
+}
+
+func NewServer(address string, logger *zap.Logger, deviceCounter deviceCounter, deviceCheckInterval time.Duration) *Server {
+	return &Server{
+		address:              address,
+		logger:               logger,
+		deviceCounter:        deviceCounter,
+		listAndWatchDoneChan: make(chan struct{}),
```

Comment:
Indeed it is a shutdown channel, I've usually thought it was a code smell to stick a context in a struct, but maybe it's fine if well documented

Reply:
Renamed to listAndWatchShutdownChan

Reply:
> I've usually thought it was a code smell to stick a context in a struct

Sticking the whole context in, kinda yes. But I think here we can capture its Done chan and use that without needing to create and close our own separate channel, since need to be able to shut down the GRPC stream (which has an embedded context) in the ListAndWatch loop later